maps-compose/src/main/java/com/google/maps/android/compose/MapApplier.kt

```diff
@@ -68,94 +68,138 @@ internal class MapApplier(
-                    this?.markerState?.dragState = DragState.END
-                }
+                decorations.findInputCallback<MarkerNode, Marker, Unit>(
+                    nodeMatchPredicate = { it.marker == marker },
+                    nodeInputCallback = {
+                        {
+                            markerState.position = it.position
+                            markerState.dragState = DragState.END
+                        }
+                    },
+                    inputHandlerCallback = { onMarkerDragEnd }
+                )?.invoke(marker)
             }
 
             override fun onMarkerDragStart(marker: Marker) {
-                with(decorations.nodeForMarker(marker)) {
-                    this?.markerState?.position = marker.position
-                    this?.markerState?.dragState = DragState.START
-                }
+                decorations.findInputCallback<MarkerNode, Marker, Unit>(
+                    nodeMatchPredicate = { it.marker == marker },
+                    nodeInputCallback = {
+                        {
+                            markerState.position = it.position
+                            markerState.dragState = DragState.START
+                        }
+                    },
+                    inputHandlerCallback = { onMarkerDragStart }
+                )?.invoke(marker)
             }
         })
         map.setInfoWindowAdapter(
             ComposeInfoWindowAdapter(
                 mapView,
-                markerNodeFinder = { decorations.nodeForMarker(it) }
+                markerNodeFinder = { marker ->
+                    decorations.firstOrNull { it is MarkerNode && it.marker == marker }
+                            as MarkerNode?
+                }
             )
         )
     }
 }
 
-private fun MutableList<MapNode>.nodeForCircle(circle: Circle): CircleNode? =
-    firstOrNull { it is CircleNode && it.circle == circle } as? CircleNode
-
-private fun MutableList<MapNode>.nodeForMarker(marker: Marker): MarkerNode? =
-    firstOrNull { it is MarkerNode && it.marker == marker } as? MarkerNode
-
-private fun MutableList<MapNode>.nodeForPolygon(polygon: Polygon): PolygonNode? =
-    firstOrNull { it is PolygonNode && it.polygon == polygon } as? PolygonNode
-
-private fun MutableList<MapNode>.nodeForPolyline(polyline: Polyline): PolylineNode? =
-    firstOrNull { it is PolylineNode && it.polyline == polyline } as? PolylineNode
-
-private fun MutableList<MapNode>.nodeForGroundOverlay(
-    groundOverlay: GroundOverlay
-): GroundOverlayNode? =
-    firstOrNull { it is GroundOverlayNode && it.groundOverlay == groundOverlay } as? GroundOverlayNode
+/**
+ * General pattern for handling input:
+ * Find the node that belongs to the clicked item.
+ * If there is none, default to the first InputHandlerNode.
+ * If there is none, don't handle.
+ */
+private inline fun <reified NodeT : MapNode, I, O> Iterable<MapNode>.findInputCallback(
```

Comment:
👌

Reply:
Honestly worried this is "too clever", but it certainly solves the problem

Reply:
It's definitely much cleaner!